File: cpp/hot100/58.组合总和.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        int n = candidates.size();
        vector<vector<int>> ans;
        vector<int> path;
        int sum = 0;
        auto dfs = [&](this auto&& dfs, int i, int sum) -> void{
            if(sum == target){
                ans.emplace_back(path);
                return;
            }
            if(sum > target || i == n) return;
            dfs(i + 1, sum);
            path.emplace_back(candidates[i]);
            dfs(i, sum + candidates[i]);
            path.pop_back();
        };
        dfs(0, 0);
        return ans;
    }
};

File: python/hot100/58.组合总和.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        n = len(candidates)
        sum = 0
        ans = []
        path = []
        def dfs(i, sum) -> None:
            if sum is target:
                ans.append(path.copy())
                return
            if sum > target or i == n:
                return 
            dfs(i + 1, sum)
            path.append(candidates[i])
            dfs(i, sum + candidates[i])
            path.pop()
        dfs(0, 0)
        return ans